Load PTIB

stephanieyurchak opened the pull request:
I added a load script for PTIB built off of the stp enrol branch. 

Decided to go with machine readable column names compared to previous cycles:
year                                   - Year
credential                          - Credential (Program) (Program)
cip                                     - CIP
age_group                         - Age Group
immigration_status           - Immigration Status
sum_of_graduates             - Sum of Graduates
sum_of_enrolments           - Sum of Enrolments
sum_of_total_enrolments  - Sum of Total Enrolments

Name of table to be loaded into decimal is PTIB_Credentials based on the table name from PSSM_2019-2020_Private Training Institutions.R (T_Private_Institutions_Credentials_Imported_2021-03).

BonnieJRobert commented on R/load-ptib.R:
```diff
@@ -0,0 +1,60 @@
+library(DBI)
+library(safepaths)
+library(config)
+library(xlsx)
```
I've always encountered a problem with Java when loading `xlsx`.  `readxl` doesn't have the same java dependency so requesting we switch.

BonnieJRobert commented on R/load-ptib.R:
```diff
@@ -0,0 +1,60 @@
+raw_data_file <- glue::glue("{ptib}/PTIB 2021 and 2022 Enrolment Data for BC Stats.xlsx")
+
+## ----- Read raw data  ----
+raw_data <- read.xlsx(raw_data_file, sheetIndex = 1, startRow = 3) %>% 
+  janitor::clean_names() %>% 
+  rename(year = calendar_year,
+         graduates = credential_1) ## fairly sure this is the case since this 
+                                   ## column is equal to total_enrolments - enrolments_not_graduated
```
```suggestion
raw_data <- readxl::read_xlsx(raw_data_file, sheet = 1, skip = 3) %>% 
  janitor::clean_names() %>% 
  rename(year = calendar_year,
         credential = credential_6,
         graduates = credential_8, 
         age_group = age_range) ## fairly sure this is the case since this 
                                   ## column is equal to total_enrolments - enrolments_not_graduated
```

BonnieJRobert commented on R/load-ptib.R:
```diff
@@ -0,0 +1,60 @@
+                  name = "PTIB_Credentials",
+                  nanoarrow::as_nanoarrow_array_stream(data))
+
+
```
```suggestion
# DBI::dbReadTable(con, "PTIB_Credentials")
```

BonnieJRobert reviewed (changes requested):
Looks great!  requesting we use `readxl` - I've generally found it more friendly.  Also, we should read back to R and do some QA before finalizing.

stephanieyurchak commented:
I updated to readxl, added dbreadtable statement to the testing code at the bottom, and switched to the new data. I had to add some cleaning code for the cip column with the new data.

lindsay-fredrick commented on R/load-ptib.R:
```diff
@@ -0,0 +1,98 @@
+         cip3 = paste(cip1, cip2, sep = "."))
+
+## Compare adjusted cip codes with the statistics canada definitions here:
+## https://www.statcan.gc.ca/en/subjects/standard/cip/2016/index
```
Is 2016 the correct version to be using here - or will new data mean the new 2021 cip codes (https://www.statcan.gc.ca/en/subjects/standard/cip/2021/index)

stephanieyurchak replied:
This is the final version of the data we will be receiving from PTIB.

lindsay-fredrick commented on R/load-ptib.R:
```diff
@@ -0,0 +1,98 @@
+                 Trusted_Connection = "True")
+
+# ---- Define schema ----
+schema <-
```
This isn't used - is it needed?

stephanieyurchak replied:
For some reason I thought it was being used by dbWriteTableArrow...

lindsay-fredrick commented on R/load-ptib.R:
```diff
@@ -0,0 +1,98 @@
+# ---- Write to decimal ----
+dbWriteTableArrow(con,
+                  name = "PTIB_Credentials",
+                  nanoarrow::as_nanoarrow_array_stream(data))
```
This is mostly just a curiousity comment for myself - is nanoarrow just much faster to load in? Seems like this is a reasonably small table that doesn't need the intermediate step (I've also never used nanoarrow, so that's neat!) - but if it really helps the bigger ones then it probably makes sense to read all the things in the same way!

stephanieyurchak replied:
It was used for STP data (which is very big). @BonnieJRobert, do you know if it was much faster?

BonnieJRobert replied:
TBH, I didn't have enough patience to compare methods on the enrolment data, also the time savings will depend on whether one is logged into Citrix or not, and current LAN speeds, too.  It makes sense to me, though, that `nanoarrow` is faster in the case where a file is very large.

lindsay-fredrick commented on R/load-ptib.R:
```diff
@@ -0,0 +1,98 @@
+dbWriteTableArrow(con,
+                  name = "PTIB_Credentials",
+                  nanoarrow::as_nanoarrow_array_stream(data))
+
```
I'm less knowledgeable here - do we need to disconnect from the database at the end?
`dbDisconnect()` ?

stephanieyurchak replied:
I will add it in

lindsay-fredrick reviewed (commented):
Looks good to me! I think my only real request would maybe be to update the way we pick up the lan path - seems like it can be added to everyone's config files. Otherwise nice work wrangling that incredibly annoying excel rounding!

stephanieyurchak commented on R/load-ptib.R:
```diff
@@ -0,0 +1,98 @@
+library(janitor)
+
+# ---- Configure LAN Paths and DB Connection ----
+lan <- get_network_path()
```
I used what Bonnie set up for load-stp-enrol. @BonnieJRobert, what do you think of using config over the safepaths route?

BonnieJRobert replied:
Yes, let's use config for everything and keep things consistent.  It'll be easier when swapping between the two shares, plus file paths are referenced differently when logged into Alder through Citrix.   I've put my config.R file in the top-level project directory on the LAN, for reference.
